### PR TITLE
refactor: extract cache policies

### LINKED ISSUE #15: Split cache policies into separate files (if possible)
Please create 3 different files for the different polling modes if possible:
auto_polling_cache_policy.ex
lazy_loading_cache_policy.ex
manual_polling_cache_policy.ex

_Originally posted by @laliconfigcat in https://github.com/configcat/elixir-sdk/pull/2#discussion_r506363541_

### LINKED ISSUE #37: RFC: New architecture to support desired features
## Background

In order to support a number of desired features that would make the SDK more extensible and closer in structure to the other language SDK's, I'm proposing a restructuring of the architecture as outlined below.

Here are the features we'd like to support:
- #10 - Allow a custom cache implementation
- #14 - Extract config logic to a ConfigFetcher module
- #15 - Split cache policies into separate files (if possible)
- #30 - Improve public interface

I've based this "desired future" largely on the current Ruby SDK as that's the one I'm most familiar with.

## Current State

- Almost all of the logic lives in the `ConfigCat` GenServer.

- The state of the GenServer includes both client settings and the currently-cached config.

- FetchPolicies are simple structs and the `FetchPolicy` module provides some helper functions that pattern match based on the type of the policy.

Something that's not clear from the tests is how this library would be used within an application, and I think this lack of clarity is what spawned #30.

Once we add a sample application (#5) and documentation (#4) this will become more clear, but here's a rough summary:

- The current tests make it seem like we have to call `ConfigCat.start_link` every time we want to request a value; however, that's not the intended use for a client application.

- Instead, a client application adds `ConfigCat` to its supervision tree as appropriate using a child spec such as `{ConfigCat, [sdk_key()]}`.

- The client can then call `ConfigCat.get_value(key, default, user)` as appropriate.

- By default, the public API in `ConfigCat` will talk to the running GenServer process using its default name (the `ConfigCat` module name).

- It is possible to provide a different name for the GenServer (as is done in the tests), and to have the public API talk to a different GenServer via an optional `client: pid` argument. These extra arguments are there mostly for the tests, or for a more complicated use case where we want to use multiple different ConfigCat configurations in the same application.  In normal circumstances, the client application doesn't need to provide those arguments.

## Design Considerations

- Since Elixir is not an object-oriented language like Ruby, we can't simply initialize a class with some instance variables that we can then call functions on.

- The current `ConfigCat` module has too many responsibilities.

- The current architecture does not easily support a pluggable cache implementation.

- Other language clients split their cache policies into separate files/classes.

- In Elixir, persistent state is normally managed by GenServers as are asynchronous periodic processes (like the auto-polling cache policy).

## Proposal

- Split `ConfigFetcher` out to its own (non-GenServer) module.

- Define an Elixir behaviour for a `ConfigCache`.

- Rename `FetchPolicy` to `CachePolicy` to match the other language SDKs; other changes as outlined below.

- Divide the rest of the application into a Supervisor with a family of child GenServers:
  - A `Client` GenServer will manage any global application state (similar to the inst vars in the Ruby `Client` class). This state will include the configured cache policy (renamed from `FetchPolicy`) and anything else necessary.

  - A `CachePolicy` GenServer: one of `CachePolicy.ManualPolling`, `CachePolicy.AutoPolling`, or `CachePolicy.LazyLoading`. Whichever of these GenServers is running will keep a reference to the cache and config fetcher as well as its own settings.  The AutoPolling GenServer will also manage the polling loop.

  - (optional) A `Cache` GenServer: we'll provide a `Cache.InMemory` GenServer as a default and add it to our supervision tree. This GenServer will manage the current configuration as its state.  More on pluggable cache implementations below.

- We might need to find a way to eliminate duplication between the CachePolicy GenServers.

- The way the calling application specifies the cache policy might change; internally, we'll turn that specification into a child_spec that we can add to our supervision tree.

- The calling application can provide a custom cache implementation as either `cache_module: SomeModule` or `cache_spec: some_child_spec`.  In the former case, we'll call functions on that module; in the latter, we'll add the cache to our supervision tree (and then call functions on the module).  This will allow callers to provide a cache (like Redis) that is already supervised by their application, or a simpler cache (like our `InMemoryCache`) that needs to be started and supervised.

- The public API stays largely as it is today: the calling application adds `ConfigCat` to its supervision tree and then calls public API functions like `ConfigCat.get_value(key, default, user)`.  This goes against what #30 is asking for, but I'm hoping that the explanation above shows that this is already a very usable (and Elixir-like) API.

This is a more complex architecture than what we have today. However, it also divides up responsibilities into smaller, more easily-testable (and pluggable) pieces. It is also structured much more like the other language SDKs.

## Roadmap

Here's how I think we can get to this new architecture incrementally:

- [x] Extract a `ConfigFetcher` module to manage the communication with the ConfigCat servers. (#40)

- [x] Add a Supervisor above the current `ConfigCat` GenServer. (#38)

- [x] Extract a `ConfigCache` behaviour and `InMemory` implementation of that GenServer.

- [x] Convert `FetchPolicy` into a family of `CachePolicy` GenServers.

### PR DESCRIPTION
**Pull request by randycoulman**
Closes #15 
Closes #37 

Splits `FetchPolicy` into a `CachePolicy` behaviour and convenience API and separate sub-modules for the 3 cache policies.  All of the fetching/caching lives in the cache policies now, and the Client module is much simpler.

This is still a draft because I want to eliminate some duplication between the various fetch policies and their tests before merging.

### CODE REVIEW

File: test/config_cat/cache_policy/auto_test.exs [hunk 1]
```diff
@@ -0,0 +1,132 @@
+defmodule ConfigCat.CachePolicy.AutoTest do
```
Comment (igorescobar) [inline test/config_cat/cache_policy/auto_test.exs:1]: It's nice that with this split the tests are also now on different files, looks much cleaner.

File: test/support/mocks.ex [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Mox.defmock(ConfigCat.MockAPI, for: HTTPoison.Base)
```
Comment (igorescobar) [inline test/support/mocks.ex:1]: 🥳

**Review by igorescobar — APPROVED** [review]
Looking good!

**Comment by randycoulman** [comment]
@igorescobar I've finished this PR now; can you give it a final 👍 . The last two commits remove a bunch of duplication from the CachePolicy implementations and their tests.

**Review by igorescobar — APPROVED** [review]
Really liked what you did here @randycoulman 🍾 

Awesome job!

**Pull request merged**